Update nvidia_devices to call into nvidia-container-runtime-hook

RenaudWasTaken opened the pull request:
This PR changes 2 things:
1. We changes the behavior of nvidia_devices to call into nvidia-container-runtime-hook rather than call into the nvidia-container-cli.
2. We fix the nvidia driver's registration which only registered two caps (gpu and nvidia)

The reason 1. is needed is because the the nvidia-container-runtime-hook operates on the nvidia container images (thorugh the [env API](https://github.com/NVIDIA/nvidia-container-runtime#environment-variables-oci-spec)) as well as some other features related to logging, debugging, ...
The nvidia-container-cli does not read the OCI spec, keeping the CLI would have the unfortunate behavior of breaking all the NVIDIA images (nvidia/cuda).

As we move towards a Device Plugin model, the topic of environment variables and how we can transition towards a better model will be one of the main discussion points.
In a future patch this hook binary will be a Device Plugin on it's own and call into the nvidia-container-cli.

Signed-off-by: Renaud Gaubert <rgaubert@nvidia.com>
cc @tiborvass 

Notes after this patch is applied, the following tests passes:
```
nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
GPU 1: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
sudo ./build/docker run -it --gpus all nvidia/cuda nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
GPU 1: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
sudo ./build/docker run -it --gpus 1 nvidia/cuda nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
sudo ./build/docker run -it --gpus device=1 nvidia/cuda nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
sudo ./build/docker run -it --gpus all,capabilities=utility debian:8-slim nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
GPU 1: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
```

**- A picture of a cute animal (not mandatory but encouraged)**
![maxresdefault](https://user-images.githubusercontent.com/2519571/55039226-973aea00-4fe0-11e9-8eb6-12c65fb92248.jpg)

Commit:
daemon: fix bug in nvidia device registration

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
Update nvidia_devices to call into nvidia-container-runtime-hook

Signed-off-by: Renaud Gaubert <rgaubert@nvidia.com>

tiborvass commented on daemon/nvidia_linux.go:
```diff
@@ -46,62 +48,83 @@ func init() {
+		Env: os.Environ(),
+	})
+
 	return nvidia.WithGPUs(opts...)(nil, nil, nil, s)
```
@RenaudWasTaken there will be two prestart hooks here, the one you added in this code, and another one that's added from containerd's contrib. Also, I wonder how much of this change should be in containerd's contrib instead.

RenaudWasTaken replied:
Thanks, for noticing! I've fixed that.

thaJeztah commented on daemon/nvidia_linux.go:
```diff
@@ -38,70 +40,90 @@ func init() {
+	m := make(map[string]string)
+	for _, s := range e {
+		p := strings.SplitN(s, "=", 2)
+		m[p[0]] = p[1]
```
This will panic if there's no `=` in `s`

RenaudWasTaken replied:
Thanks for noticing! Fixed

thaJeztah reviewed (commented):
Bit confused by this change as it seems to be a 180-degrees change of the previous design;

> The nvidia-container-cli does not read the OCI spec

Is there a reason it doesn't? (i.e. if it would, wouldn't it be able to obtain all the required information?)

tiborvass commented:
@thaJeztah Why do you think it's a 180 change? It's actually implementing correctly what we wanted in the first place, but incorrectly tested it.

thaJeztah commented:
I thought the intent was to get rid of the environment-variables as an intermediate step (as those IIUC were used as a stop-gap solution until there was an official way to pass options to the nvidia driver). The driver could still look for environment variables (if present), but just wondering if continuing to use them as the way to pass these is the right thing to do.

RenaudWasTaken commented:
> Bit confused by this change as it seems to be a 180-degrees change of the previous design;
> > The nvidia-container-cli does not read the OCI spec
> Is there a reason it doesn't? (i.e. if it would, wouldn't it be able to obtain all the required information?)

nvidia-container-runtime-hook is really just a go wrapper around the nvidia-container-cli
    - This go wrapper is available here: https://github.com/NVIDIA/nvidia-container-runtime/tree/master/hook/nvidia-container-runtime-hook
     - What this go wrapper does is parse the OCI spec for some of the fields that are present in the container images [(NVIDIA_REQUIRE_CUDA, NVIDIA_CAPABILITIES, ...)](https://github.com/NVIDIA/nvidia-container-runtime#environment-variables-oci-spec)
     - It also takes care of reading the nvidia config file for things like logging, debug level, libs rootfs (in the special case of the driver container)

I didn't realize that containerd was calling directly into the CLI and doing its own parsing based only on the CLI argument given. This is problematic because the images have some critical business logic (e.g: This image runs only if your driver supports CUDA 10).

Calling into the hook also helps reduce the amount of nvidia specific code inside docker as well as reduce the amount of bugs as this is code running in production for quite some time.
Finally, In the desired model, the code part of the hook will most likely be the code of docker device plugin.

Does this help answer some of the problems you see?

RenaudWasTaken commented:
> I thought the intent was to get rid of the environment-variables as an intermediate step (as those IIUC were used as a stop-gap solution until there was an official way to pass options to the nvidia driver). The driver could still look for environment variables (if present), but just wondering if continuing to use them as the way to pass these is the right thing to do.

@thaJeztah that's definitely something we want to move away from as we roll in the concept of docker device plugin. We also want a smooth transition from nvidia-docker2 (if possible not break the images) and drop support for the env variables over time.

I think one of the prominent intent of this effort for the 19.03 release is to introduce the concept of --gpus while at the same time start the deprecation process for the nvidia-docker2 packages.

Does this sound like a reasonable course of action?

tiborvass reviewed (approved):
LGTM

thaJeztah commented on daemon/nvidia_linux.go:
```diff
@@ -38,70 +40,96 @@ func init() {
+			nvidiaHook,
+			"prestart",
+		},
+		Env: os.Environ(),
```
Does this need to get `os.Environ()` ?

RenaudWasTaken replied:
Yep, because the hook needs to lookup the nvidia-container-cli which it will need to PATH variable set.

thaJeztah replied:
Ah, right; wasn't sure how the hook was started; if it's `exec.Cmd`, it would already use the current environment;

https://golang.org/pkg/os/exec/#Cmd

```
        // Env specifies the environment of the process.
        // Each entry is of the form "key=value".
        // If Env is nil, the new process uses the current process's
        // environment.
        // If Env contains duplicate environment keys, only the last
        // value in the slice for each duplicate key is used.
        Env []string
```

thaJeztah commented on daemon/nvidia_linux.go:
```diff
@@ -38,70 +40,96 @@ func init() {
 
 func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
-	var opts []nvidia.Opts
+	envMap, err := getEnvMap(s.Process.Env)
```
The conversion to a map is only to make sure we "overwrite" the previous value, correct?

Given that we're only appending new values (and don't look at existing values), I don't think there's a need to do this conversion, as `FOOBAR=one FOOBAR=two` is equivalent to `FOOBAR=two` (latter value overwrites former value);

```patch
diff --git a/daemon/nvidia_linux.go b/daemon/nvidia_linux.go
index 6756187649..86e4c0d006 100644
--- a/daemon/nvidia_linux.go
+++ b/daemon/nvidia_linux.go
@@ -1,7 +1,6 @@
 package daemon
 
 import (
-       "fmt"
        "os"
        "os/exec"
        "strconv"
@@ -47,22 +46,17 @@ func init() {
 }
 
 func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
-       envMap, err := getEnvMap(s.Process.Env)
-       if err != nil {
-               return err
-       }
-
        req := dev.req
        if req.Count != 0 && len(req.DeviceIDs) > 0 {
                return errConflictCountDeviceIDs
        }
 
        if len(req.DeviceIDs) > 0 {
-               envMap["NVIDIA_VISIBLE_DEVICES"] = strings.Join(req.DeviceIDs, ",")
+               s.Process.Env = append(s.Process.Env, "NVIDIA_VISIBLE_DEVICES="+strings.Join(req.DeviceIDs, ","))
        } else if req.Count > 0 {
-               envMap["NVIDIA_VISIBLE_DEVICES"] = countToDevices(req.Count)
+               s.Process.Env = append(s.Process.Env, "NVIDIA_VISIBLE_DEVICES="+countToDevices(req.Count))
        } else if req.Count < 0 {
-               envMap["NVIDIA_VISIBLE_DEVICES"] = "all"
+               s.Process.Env = append(s.Process.Env, "NVIDIA_VISIBLE_DEVICES=all")
        }
 
        var nvidiaCaps []string
@@ -78,11 +72,9 @@ func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
        }
 
        if nvidiaCaps != nil {
-               envMap["NVIDIA_DRIVER_CAPABILITIES"] = strings.Join(nvidiaCaps, ",")
+               s.Process.Env = append(s.Process.Env, strings.Join(nvidiaCaps, ","))
        }
 
-       s.Process.Env = envMapToArray(envMap)
-
        path, err := exec.LookPath(nvidiaHook)
        if err != nil {
                return err
@@ -111,25 +103,3 @@ func countToDevices(count int) string {
        }
        return strings.Join(devices, ",")
 }
-
-func getEnvMap(e []string) (map[string]string, error) {
-       m := make(map[string]string)
-       for _, s := range e {
-               p := strings.SplitN(s, "=", 2)
-               if len(p) != 2 {
-                       return nil, fmt.Errorf("Failed to split environment variable `%s`", p)
-               }
-               m[p[0]] = p[1]
-       }
-       return m, nil
-}
-
-func envMapToArray(m map[string]string) []string {
-       env := make([]string, len(m))
-       i := 0
-       for k, v := range m {
-               env[i] = fmt.Sprintf("%s=%s", k, v)
-               i++
-       }
-       return env
-}
```

RenaudWasTaken replied:
This is a lot better!

Thanks!

RenaudWasTaken commented on daemon/nvidia_linux.go:
```diff
@@ -38,70 +40,96 @@ func init() {
+	for _, s := range e {
+		p := strings.SplitN(s, "=", 2)
+		if len(p) != 2 {
+			return nil, fmt.Errorf("Failed to split environment variable `%s`", p)
```
This is resolved thanks to your later comment!

RenaudWasTaken commented:
@thaJeztah I've addressed your comments and made sure the tests passed again:
```
nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
GPU 1: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
sudo ./build/docker run -it --gpus all nvidia/cuda nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
GPU 1: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
sudo ./build/docker run -it --gpus 1 nvidia/cuda nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
sudo ./build/docker run -it --gpus device=1 nvidia/cuda nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
sudo ./build/docker run -it --gpus all,capabilities=utility debian:8-slim nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
GPU 1: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
sudo ./build/docker run -it -e "NVIDIA_REQUIRE_CUDA=cuda>=10" --gpus all nvidia/cuda nvidia-smi -L
GPU 0: GeForce GTX 1070 (UUID: GPU-604ac76c-d9cf-fef3-62e9-d92044ab6e52)
GPU 1: GeForce GTX 1070 (UUID: GPU-6a2e4b48-c3ac-f34e-df41-88c04acb0709)
```

I'll open a PR next week which adds some level of mock testing. With @tiborvass we've also talked about having a weekly E2E test pipeline setup.

thaJeztah commented:
Will the `NVIDIA_REQUIRE_CUDA=cuda>=10` eventually be moved to be an `option` after `--gpu` ?

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
> I'll open a PR next week which adds some level of mock testing. With @tiborvass we've also talked about having a weekly E2E test pipeline setup.

Also happy to test here locally if you send me a nice machine 😂

thaJeztah commented:
@tiborvass PTAL
